File: shared/src/commonMain/kotlin/org/athletica/crm/api/schemas/settings/DashboardSettings.kt
package org.athletica.crm.api.schemas.settings

import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonClassDiscriminator
import org.athletica.crm.core.entityids.DashboardWidgetId

/**
 * Виджет главной страницы — сущность с собственной идентичностью и настройками.
 * Видимостью и порядком виджет НЕ владеет: это concern размещения ([DashboardSettings.layout]),
 * уровнем выше виджета. Sealed-иерархия сериализуется полиморфно с дискриминатором `type`.
 */
@OptIn(ExperimentalSerializationApi::class)
@Serializable
@JsonClassDiscriminator("type")
sealed interface DashboardWidget {
    /** Идентификатор экземпляра. Позволяет иметь несколько виджетов одного типа. */
    val id: DashboardWidgetId

    /**
     * Пользовательский заголовок виджета на экране. `null` или пустая строка — использовать
     * локализованное имя типа виджета по умолчанию. Непустое значение — переименование.
     */
    val title: String?

    /** «Занятия сегодня». Кроме заголовка, настроек пока нет. */
    @Serializable
    @SerialName("sessions")
    data class Sessions(
        override val id: DashboardWidgetId,
        override val title: String? = null,
    ) : DashboardWidget

    /** «Должники»: [limit] — максимум строк в виджете. */
    @Serializable
    @SerialName("debtors")
    data class Debtors(
        override val id: DashboardWidgetId,
        override val title: String? = null,
        val limit: Int = DEFAULT_LIMIT,
    ) : DashboardWidget {
        companion object {
            const val DEFAULT_LIMIT = 10
        }
    }

    /** «Дни рождения»: [window] — окно отбора именинников, [limit] — максимум строк. */
    @Serializable
    @SerialName("birthdays")
    data class Birthdays(
        override val id: DashboardWidgetId,
        override val title: String? = null,
        val window: BirthdayWindow = BirthdayWindow.TODAY,
        val limit: Int = DEFAULT_LIMIT,
    ) : DashboardWidget {
        companion object {
            const val DEFAULT_LIMIT = 10
        }
    }
}

/** Окно отбора именинников относительно текущей даты. */
@Serializable
enum class BirthdayWindow {
    /** Только сегодня. */
    TODAY,

    /** Только завтра. */
    TOMORROW,

    /** Ближайшая неделя (сегодня — сегодня+7). */
    WEEK,
}

/**
 * Настройки дашборда главной страницы.
 * [widgets] — пул сконфигурированных экземпляров виджетов (вся конфигурация живёт здесь).
 * [layout] — идентификаторы показанных виджетов в порядке отображения; видимость =
 * членство в [layout], порядок = порядок в [layout]. Скрытый виджет остаётся в [widgets]
 * (сохраняет настройки), но отсутствует в [layout].
 */
@Serializable
data class DashboardSettings(
    val widgets: List<DashboardWidget> = emptyList(),
    val layout: List<DashboardWidgetId> = emptyList(),
) {
    /** Видимые виджеты в порядке [layout]. Идентификаторы без виджета в пуле игнорируются. */
    fun orderedVisible(): List<DashboardWidget> {
        val byId = widgets.associateBy { it.id }
        return layout.mapNotNull { byId[it] }
    }

    /** Скрытые виджеты: присутствуют в [widgets], но отсутствуют в [layout]. */
    fun hidden(): List<DashboardWidget> {
        val shown = layout.toSet()
        return widgets.filterNot { it.id in shown }
    }

    companion object {
        /**
         * Канонический дефолтный набор: по одному экземпляру каждого типа виджета,
         * все показаны в порядке объявления. Идентификаторы генерируются при каждом вызове.
         */
        fun default(): DashboardSettings {
            val widgets =
                listOf(
                    DashboardWidget.Sessions(DashboardWidgetId.new()),
                    DashboardWidget.Debtors(DashboardWidgetId.new()),
                    DashboardWidget.Birthdays(DashboardWidgetId.new()),
                )
            return DashboardSettings(widgets = widgets, layout = widgets.map { it.id })
        }
    }
}

/**
 * Дополняет настройки недостающими дефолтными виджетами. Пустой пул заменяется полным
 * [DashboardSettings.default]; иначе для каждого типа из дефолтного набора, отсутствующего
 * в пуле, добавляется его экземпляр в конец [DashboardSettings.widgets] и [DashboardSettings.layout].
 * Покрывает нового пользователя (пустые настройки) и типы виджетов, добавленные в код позже.
 */
fun DashboardSettings.withDefaults(): DashboardSettings {
    if (widgets.isEmpty()) {
        return DashboardSettings.default()
    }
    val presentTypes = widgets.map { it::class }.toSet()
    val missing = DashboardSettings.default().widgets.filterNot { it::class in presentTypes }
    if (missing.isEmpty()) {
        return this
    }
    return copy(widgets = widgets + missing, layout = layout + missing.map { it.id })
}

File: web/src/home/dashboardSettings.ts
import {
  DashboardWidgetIdSchema,
  LocalDateSchema,
  type BirthdayWindow,
  type ClientListRequest,
  type DashboardSettings,
  type DashboardWidget,
  type DashboardWidgetBirthdays,
  type DashboardWidgetDebtors,
  type DashboardWidgetId,
  type DateRange,
  type LocalDate,
} from "@/api/generated/contracts";
import type { ClientListSearch } from "@/clients/clientListSearch";
import type { PlainMessageKey } from "@/i18n/context";

/** Значение количества элементов в виджете по умолчанию (паритет с KMP-клиентом). */
export const DEFAULT_WIDGET_LIMIT = 10;

/** Минимальное и максимальное значение лимита виджета. */
export const MIN_WIDGET_LIMIT = 1;
export const MAX_WIDGET_LIMIT = 50;

/** Типы виджетов главной в порядке, в котором `default()` создаёт их в KMP-клиенте. */
const WIDGET_TYPES = ["sessions", "debtors", "birthdays"] as const;

/** Новый идентификатор виджета. */
function newWidgetId(): DashboardWidgetId {
  return DashboardWidgetIdSchema.parse(crypto.randomUUID());
}

/** Виджет типа [type] с настройками по умолчанию и новым идентификатором. */
function defaultWidgetOf(type: DashboardWidget["type"]): DashboardWidget {
  switch (type) {
    case "sessions":
      return { type: "sessions", id: newWidgetId(), title: null };
    case "debtors":
      return { type: "debtors", id: newWidgetId(), title: null, limit: DEFAULT_WIDGET_LIMIT };
    case "birthdays":
      return {
        type: "birthdays",
        id: newWidgetId(),
        title: null,
        window: "TODAY",
        limit: DEFAULT_WIDGET_LIMIT,
      };
  }
}

/** Настройки дашборда по умолчанию: по одному виджету каждого типа, все видимы. */
export function defaultDashboardSettings(): DashboardSettings {
  const widgets = WIDGET_TYPES.map(defaultWidgetOf);
  return { widgets, layout: widgets.map((widget) => widget.id) };
}

/**
 * Настройки [settings] с добавленными недостающими типами виджетов (паритет с
 * `DashboardSettings.withDefaults()` KMP-клиента): пустой пул — полный набор по умолчанию,
 * иначе в конец пула и раскладки добавляется по одному экземпляру каждого отсутствующего типа.
 */
export function withDashboardDefaults(settings: DashboardSettings): DashboardSettings {
  if (settings.widgets.length === 0) {
    return defaultDashboardSettings();
  }
  const present = new Set(settings.widgets.map((widget) => widget.type));
  const missing = WIDGET_TYPES.filter((type) => !present.has(type)).map(defaultWidgetOf);
  if (missing.length === 0) {
    return settings;
  }
  return {
    widgets: [...settings.widgets, ...missing],
    layout: [...settings.layout, ...missing.map((widget) => widget.id)],
  };
}

/** Видимые виджеты [settings] в порядке раскладки. */
export function orderedVisible(settings: DashboardSettings): readonly DashboardWidget[] {
  const byId = new Map(settings.widgets.map((widget) => [widget.id, widget] as const));
  return settings.layout.flatMap((id) => {
    const widget = byId.get(id);
    return widget === undefined ? [] : [widget];
  });
}

/** Скрытые виджеты [settings]: есть в пуле, но не в раскладке. */
export function hiddenWidgets(settings: DashboardSettings): readonly DashboardWidget[] {
  const visible = new Set<DashboardWidgetId>(settings.layout);
  return settings.widgets.filter((widget) => !visible.has(widget.id));
}

/** Ключ локализованного заголовка по умолчанию для типа виджета [type]. */
export function defaultTitleKey(type: DashboardWidget["type"]): PlainMessageKey {
  switch (type) {
    case "sessions":
      return "home.todaySessionsTitle";
    case "debtors":
      return "home.debtorsTitle";
    case "birthdays":
      return "home.birthdaysTitle";
  }
}

/** Заголовок виджета [widget]: свой, если не пуст, иначе локализованный по умолчанию. */
export function resolveTitle(t: (key: PlainMessageKey) => string, widget: DashboardWidget): string {
  const title = widget.title;
  return title !== null && title.trim() !== "" ? title : t(defaultTitleKey(widget.type));
}

/** Заменяет в [settings] виджет с тем же идентификатором, что у [updated]. */
export function replaceWidget(
  settings: DashboardSettings,
  updated: DashboardWidget,
): DashboardSettings {
  return {
    ...settings,
    widgets: settings.widgets.map((widget) => (widget.id === updated.id ? updated : widget)),
  };
}

/** Скрывает виджет [id]: убирает из раскладки, пул не меняется — настройки виджета сохраняются. */
export function hideWidget(settings: DashboardSettings, id: DashboardWidgetId): DashboardSettings {
  return { ...settings, layout: settings.layout.filter((candidate) => candidate !== id) };
}

/** Показывает виджет [id]: добавляет в конец раскладки, если его там ещё нет. */
export function showWidget(settings: DashboardSettings, id: DashboardWidgetId): DashboardSettings {
  if (settings.layout.includes(id)) {
    return settings;
  }
  return { ...settings, layout: [...settings.layout, id] };
}

/** Переставляет видимый виджет с позиции [index] на [index + shift] в раскладке. */
export function moveVisibleWidget(
  settings: DashboardSettings,
  index: number,
  shift: -1 | 1,
): DashboardSettings {
  const target = index + shift;
  if (target < 0 || target >= settings.layout.length) {
    return settings;
  }
  const layout = [...settings.layout];
  const moved = layout.splice(index, 1)[0];
  if (moved === undefined) {
    return settings;
  }
  layout.splice(target, 0, moved);
  return { ...settings, layout };
}

/** Значение лимита, ограниченное диапазоном `[MIN_WIDGET_LIMIT, MAX_WIDGET_LIMIT]`. */
export function clampWidgetLimit(value: number): number {
  return Math.min(MAX_WIDGET_LIMIT, Math.max(MIN_WIDGET_LIMIT, value));
}

/** Запрос списка клиентов для виджета должников [widget]. */
export function debtorsRequest(widget: DashboardWidgetDebtors): ClientListRequest {
  return { hasDebt: true, limit: widget.limit };
}

/** Запрос списка клиентов для виджета дней рождения [widget] от дня [today]. */
export function birthdaysRequest(
  widget: DashboardWidgetBirthdays,
  today: LocalDate,
): ClientListRequest {
  return { birthday: birthdayRange(widget.window, today), limit: widget.limit };
}

/**
 * Диапазон дней рождения для окна [window] от дня [today] (буквальный порт
 * `BirthdayWindow.dateRange()` KMP-клиента, включая `WEEK` = `[today, today + 7]`).
 */
function birthdayRange(window: BirthdayWindow, today: LocalDate): DateRange {
  switch (window) {
    case "TODAY":
      return { from: today, to: today };
    case "TOMORROW": {
      const tomorrow = addDaysUtc(today, 1);
      return { from: tomorrow, to: tomorrow };
    }
    case "WEEK":
      return { from: today, to: addDaysUtc(today, 7) };
  }
}

/** Дата [date] плюс [days] дней, без сдвига по часовому поясу. */
function addDaysUtc(date: LocalDate, days: number): LocalDate {
  const [year = 1970, month = 1, day = 1] = date.split("-").map(Number);
  const shifted = new Date(Date.UTC(year, month - 1, day) + days * 86_400_000);
  return LocalDateSchema.parse(shifted.toISOString().slice(0, 10));
}

/** Год даты [date] вида `2025-01-31`. */
export function yearOf(date: LocalDate): number {
  return Number(date.slice(0, 4));
}

/** Search-параметры страницы клиентов для «Показать всех» виджета должников. */
export function debtorsShowAllSearch(): ClientListSearch {
  return { debt: true };
}

/** Search-параметры страницы клиентов для «Показать всех» виджета дней рождения. */
export function birthdaysShowAllSearch(window: BirthdayWindow): ClientListSearch {
  switch (window) {
    case "TODAY":
      return { birthday: "today" };
    case "TOMORROW":
      return { birthday: "tomorrow" };
    case "WEEK":
      return { birthday: "week" };
  }
}
